### PR TITLE
Create extendable HeaderRegistry 

### PR DESCRIPTION
**Pull request by lumjjb**
Create a way to extend the functionality with custom header checking without adding dependencies into the main library, as discussed in #146 

Current implementation only takes in `JWSEHeaderParameter`, but it can be extendable to any kind of header parameter (like JWK) if we take in a set of validation functions that tells us if a header can be ovewritten, and how to reference the `check_fn`. Would like to get some feedback on this.

Side note: Took the opportunity to merge the HeaderParameter types of JWE and JWS (according to RFC 7515 - 9.1).

### CODE REVIEW

**Comment by lumjjb** [comment]
For local try-out, i used:

```
# Test check 1

# Check function takes in a JWS and checks if it's custom1 header == 'custom-text'
def check_custom1(jwobj):
    return jwobj.jose_header['custom1'] == 'custom-text'

Token = jwt.JWT(header={"alg": "RS256", 'custom1':'custom-text'}, claims={"claim1": "claim1value"})
Token.make_signed_token(key)
Token.token.header_registry['custom1'] = JWSEHeaderParameter('Custom header 1', False, True, check_custom1)

assert (Token.token.header_registry.check_header('custom1', Token.token))

Token = jwt.JWT(header={"alg": "RS256", 'custom1':'not-custom-text'}, claims={"claim1": "claim1value"})
Token.make_signed_token(key)
Token.token.header_registry['custom1'] = JWSEHeaderParameter('Custom header 1', False, True, check_custom1)

assert (not Token.token.header_registry.check_header('custom1', Token.token))
```

File: jwcrypto/jwe.py [hunk 1]
```diff
@@ -89,6 +90,7 @@ def __init__(self, plaintext=None, protected=None, unprotected=None,
         self._allowed_algs = None
         self.objects = dict()
         self.plaintext = None
+        self.header_registry = JWSEHeaderRegistry(DefaultJWEHeaderRegistry)
```
Comment (simo5) [inline jwcrypto/jwe.py]: Why the rewrap here ?

Comment (simo5) [reply jwcrypto/jwe.py]: Or maybe why DefaultJWEHeaderRegistry is a JWSEHeaderRegistry() itself instead of being just a dict ?

**Comment by simo5** [comment]
It would be nice to have a function or init() argument to allow to create a JWS/JWE and pass in custom headers to be accepted.
Also at least one unit test per JWS and JWE with a custom header.

File: jwcrypto/jwe.py (hunk 1 above)
Comment (lumjjb) [reply jwcrypto/jwe.py]: One reason I thought about this was in the current use, the header functions would have to be registered separately on each instantiation. I was thinking in the case where we had in mind to add functionality for a particular header throughout the program, in that case, we would add the header to the `DefaultRegistry`, and every object `JWS/JWE` created from then on would have that functionality built it.

**Comment by lumjjb** [comment]
> It would be nice to have a function or init() argument to allow to create a JWS/JWE and pass in custom headers to be accepted.

Got it, will add an init argument here.

> Also at least one unit test per JWS and JWE with a custom header.

Gotcha! Btw - i had a quick question about the naming of the test variables, are these particularly named to follow any examples in the spec? Or are they just arbitrary and I can create a new `A8_example`?

File: jwcrypto/jwe.py (hunk 1 above)
Comment (simo5) [reply jwcrypto/jwe.py]: uhmm, changing the DefaultJWEHeaderRegistry would make everything brittle, because i would change how every caller works, including callers in other libraries.
Please do not do that.

By allowing to pass a new registry via an init parameter in the JWE you can create your own registry once, and then always pass that one into your own controlled JWEs.

This allow the "cost savings" you want without messing up static global objects.

so I think let's get rid of DefaultJWEHeaderRegistry, and maintain only the dictionary, and move initialization in JWE/JWS constructors, including the ability to use a passed in Registry.

**Comment by simo5** [comment]
They are names after the RFC Appendix, if you need to add new tests not covered in the RFC Appendix please chose a different prefix.

File: jwcrypto/jwe.py (hunk 1 above)
Comment (lumjjb) [reply jwcrypto/jwe.py]: Yea - that makes sense. I will rename it back to `JWEHeaderRegistry` as well then so it doesn't break any libraries referencing that particular variable.

File: jwcrypto/common.py [hunk 2]
```diff
@@ -104,3 +106,83 @@ class InvalidJWEKeyLength(JWException):
...
+# HeaderParameters are for both JWS and JWE
+JWSEHeaderParameter = namedtuple('Parameter',
+                                'description mustprotect supported check_fn')
+
+class JWSEHeaderRegistry:
+    def __init__(self, init_registry=None):
+        if init_registry:
+            if isinstance(init_registry, dict):
+                self._registry = copy.deepcopy(init_registry)
+            elif isinstance(init_registry, JWSEHeaderRegistry):
```
Comment (simo5) [inline jwcrypto/common.py]: I think we can drop this now ?

File: jwcrypto/common.py [hunk 3]
```diff
@@ -104,3 +106,83 @@ class InvalidJWEKeyLength(JWException):
...
+
+
+# JWSE Header Registry definitions
+
+# RFC 7515 - 9.1: JSON Web Signature and Encryption Header Parameters Registry
+# HeaderParameters are for both JWS and JWE
+JWSEHeaderParameter = namedtuple('Parameter',
+                                'description mustprotect supported check_fn')
+
+class JWSEHeaderRegistry:
```
Comment (simo5) [inline jwcrypto/common.py]: shouldn't this simply extend 'dict' so that you do not ahve to redefine a dict-like interface, but only re-implement the __setitem__/__delitem__ calls ?

File: jwcrypto/tests.py [hunk 4]
```diff
@@ -679,6 +694,27 @@ def test_E(self):
...
 
+    def test_customhdr_jwe(self):
+        s = jws.JWS(A6_example['payload'])
+        s.deserialize(customhdr_jws_example, A6_example['key2'])
+
+        def check_custom1(jwobj):
+            return jwobj.jose_header['custom1'] == 'custom_val'
+
+        s.header_registry['custom1'] = JWSEHeaderParameter('Custom header 1', False, True, check_custom1)
+        assert (s.header_registry.check_header('custom1', s))
```
Comment (simo5) [inline jwcrypto/tests.py]: For this and following tests ... they kinda work as a sort of unit test but are not testing the integration into the JWS and JWE machinery.
I would expect to see tests that try to create/verify a JWS/JWE and see that with the incorrect/missing custom header we get an exception.

File: jwcrypto/jwe.py [hunk 5]
```diff
@@ -91,7 +92,10 @@ def __init__(self, plaintext=None, protected=None, unprotected=None,
         self._allowed_algs = None
         self.objects = dict()
         self.plaintext = None
-        self.header_registry = JWSEHeaderRegistry(header_registry)
+        if header_registry:
+            self.header_registry = JWSEHeaderRegistry(header_registry)
+        else:
+            self.header_registry = JWSEHeaderRegistry(JWEHeaderRegistry)
```
Comment (simo5) [inline jwcrypto/jwe.py]: Why this change ?

File: jwcrypto/common.py (hunk 2 above)
Comment (lumjjb) [reply jwcrypto/common.py]: Yea we can drop this now.

File: jwcrypto/common.py (hunk 3 above)
Comment (lumjjb) [reply jwcrypto/common.py]: I was hesitant about extending the `dict` class directly due to some dangers of weird built-in implementation specific issues. One of the few were:

http://www.kr41.net/2016/03-23-dont_inherit_python_builtin_dict_type.html

What do you think about this? I will defer that decision to you, I am okay with either but personal preference of not extending the `dict` class directly.

File: jwcrypto/tests.py (hunk 4 above)
Comment (lumjjb) [reply jwcrypto/tests.py]: I didn't integrate the `check_header` into the actual signature verification and decryption process. But based on this comment, I think that's something that you'd want yea?

File: jwcrypto/common.py (hunk 3 above)
Comment (simo5) [reply jwcrypto/common.py]: Uhmm  ... interesting read.
Maybe use the Mutable thing instead ?
We already import namedtuple from collections, so that is ok to have as a dependency I guess.

File: jwcrypto/tests.py (hunk 4 above)
Comment (simo5) [reply jwcrypto/tests.py]: Yea, that's the way people would expect tohave it working I think.

File: jwcrypto/common.py (hunk 3 above)
Comment (lumjjb) [reply jwcrypto/common.py]: Ok sg, i'll use that then.

File: jwcrypto/tests.py (hunk 4 above)
Comment (lumjjb) [reply jwcrypto/tests.py]: Alright, let me add that.

File: jwcrypto/jwe.py (hunk 5 above)
Comment (lumjjb) [reply jwcrypto/jwe.py]: `pep8` complained about a dictionary as a default value... Since it's mutable, and unfortunately, there isn't a frozen version of dictionary that it recognizes afaik.

**Comment by lumjjb** [comment]
Made the changes. Let me know if all is good and i'll squash.

File: jwcrypto/jws.py [hunk 6]
```diff
@@ -264,7 +264,12 @@ def _verify(self, alg, key, payload, signature, protected, header=None):
                 raise InvalidJWSSignature('Invalid Unprotected header')
 
         # Merge and check (critical) headers
-        self._merge_check_headers(p, header)
+        chk_hdrs = self._merge_check_headers(p, header)
+        for hdr in chk_hdrs:
```
Comment (simo5) [inline jwcrypto/jws.py]: I would expect headers checks to happen inside the function named _merge_check_headers() ?

File: jwcrypto/tests.py [hunk 7]
```diff
@@ -695,27 +695,26 @@ def test_E(self):
             s.verify(None)
 
     def test_customhdr_jws(self):
+        # Test pass header check
         s = jws.JWS(A6_example['payload'])
```
Comment (simo5) [inline jwcrypto/tests.py]: I would expect JWS to be initialized with the dictionary that defines a custom header check ?

File: jwcrypto/tests.py [hunk 8]
```diff
@@ -988,25 +987,23 @@ def test_JWE_Issue_136(self):
 
     def test_customhdr_jwe(self):
         e = jwe.JWE()
```
Comment (simo5) [inline jwcrypto/tests.py]: as well here JWE init should get the custom header registry ?

File: jwcrypto/jwe.py [hunk 9]
```diff
@@ -492,7 +506,7 @@ def payload(self):
 
     @property
     def jose_header(self):
-        jh = self._get_jose_header()
+        jh = self._get_jose_header(self.objects.get('header'))
```
Comment (simo5) [inline jwcrypto/jwe.py]: This is a good catch, but would be nice to have it in a separate PR I can push right away,

Comment (lumjjb) [reply jwcrypto/jwe.py]: Ok sounds good, I will create a separate PR for this!

File: jwcrypto/tests.py (hunk 7 above)
Comment (lumjjb) [reply jwcrypto/tests.py]: Got it, i added the header later on, but agree would be fitting to have it as a custom registry.

File: jwcrypto/jws.py (hunk 6 above)
Comment (lumjjb) [reply jwcrypto/jws.py]: yea sounds good, will try that.

Comment (lumjjb) [reply jwcrypto/jws.py]: This doesn't work well when I tried it. What ends up happening is that like in the header extension check function that i created in the test, which uses `jose_header`, it calls `self._merge_check_headers`, which ends up in a recursive loops.

Comment (simo5) [reply jwcrypto/jws.py]: merge and check used to be two separate functions,
let's leave the code as you have it now with a FIXME comment, and I'll take care of restructuring at a later time.

**Comment by lumjjb** [comment]
- Made the req. changes 
- made it so that only allow adding to reg
- added test for dup headers

File: jwcrypto/jwe.py [hunk 10]
```diff
@@ -354,6 +363,11 @@ def _decrypt(self, key, ppe):
         # TODO: allow caller to specify list of headers it understands
```
Comment (simo5) [inline jwcrypto/jwe.py]: Time to drop this TODO :-D

File: jwcrypto/jws.py [hunk 11]
```diff
@@ -213,6 +211,7 @@ def is_valid(self):
         return self.objects.get('valid', False)
 
     # TODO: allow caller to specify list of headers it understands
```
Comment (simo5) [inline jwcrypto/jws.py]: and this one too

**Comment by simo5** [comment]
Thanks a lot for this contribution!
